src/s2client-fsharp/Sc2Game.fs

```diff
@@ -91,86 +103,114 @@ module Sc2Game =
-                | Computer _, _ -> None
-                | _ -> failwithf "Expected playerId when participant or observer but not when computer. %A" (part,playerId)
-            )
+                |Participant (instance, _, bot), Some playerId ->
+                    PlayerData.Create playerId instance bot |> Ok |> Some
+                    //(playerId, instance, bot) :: state
+                |Observer (instance, bot), Some playerId ->
+                    PlayerData.Create playerId instance (fun data -> bot data; []) |> Ok |> Some
+                |Computer _, _ -> None
+                | _ -> sprintf "Expected playerId when participant or observer but not when computer. %A" (part,playerId) |> ConfigError |> Error |> Some
+            ) |> Result.listFold
 
         // Get the static gameInfo
-        for (playerId, instance, bot) in relevantPlayers do
-            let! gameInfo = Instance.getGameInfo instance
-            let state = getState playerId
-            updateState playerId { state with GameInfo = gameInfo }
-
-        // Game loop
-        while true do
-            for (playerId, instance, bot) in relevantPlayers do
-                let! obs = Instance.getObservation false instance
-                // TODO: Higher level support, GetUnits -> Self -> StartLocation
-                let lastState = getState playerId
-                let state = 
-                    { lastState with 
-                        NewObservation = obs
-                        LastObservation =
-                            if not (isNull lastState.NewObservation) then Some lastState.NewObservation
-                            else None }
-                
-                let actions = bot state
-                if not gameSettings.Realtime then
-                    do! Instance.doStep gameSettings.StepSize instance
-
-                updateState playerId { state with LastActions = actions }
-
-            // Execute actions
-            for (playerId, instance, bot) in relevantPlayers do
-                let lastState = getState playerId
-                do! Instance.doActions lastState.LastActions instance |> Async.Ignore
-        }
\ No newline at end of file
+        let getStaticGameInfo =
+            List.map (fun (player:PlayerData) -> 
+                Instance.getGameInfo player.Instance
+                |> Result.mapAsyncInput (fun gi -> player, gi)
+            ) >> Async.Parallel >> Async.RunSynchronously >> List.ofArray >> Result.listFold
+            
+        let getInitialGameState =
+            List.map (fun (player:PlayerData, gameInfo:ResponseGameInfo) ->
+                Instance.getObservation false player.Instance
+                |> Result.mapAsyncInput (fun obs -> player, GameState.InitialState player.PlayerId obs gameInfo)
+            ) >> Async.Parallel >> Async.RunSynchronously >> List.ofArray >> Result.listFold
+
+        let rec gameLoop playersResult =
```

Comment:
Maybe its just me but I feel the `while` loop is a lot more readable. Personally I feel like mutable state is acceptable at this level